Documentation/maxpasswd.1
.TH MAXPASSWD "1" "August 2024" "maxpasswd  " "User Commands"
.SH NAME
maxpasswd
.SH SYNOPSIS
.B maxpasswd
[\fI\,-h|--help\/\fR] [\fI\,-i|--interactive\/\fR] [\fI\,-d|--decrypt\/\fR] [\fI\,path\/\fR] \fI\,password\/\fR
.SH DESCRIPTION
Encrypt a MaxScale plaintext password using the encryption key in the key file
\&'.secrets'. The key file may be generated using the 'maxkeys'\-utility.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display this help.
.TP
\fB\-d\fR, \fB\-\-decrypt\fR
Decrypt an encrypted password instead.
.TP
\fB\-i\fR, \fB\-\-interactive\fR  \- If maxpasswd is reading from a pipe, it will read a line and
use that as the password.
.TP
\- If maxpasswd is connected to a terminal console, it will prompt
for the password.
.TP
If '\-i' is specified, a single argument is assumed to be the path
and two arguments is treated like an error.
.TP
path
The key file directory (default: '/home/markusjm/build\-develop/data/lib/maxscale')
.TP
password
The password to encrypt or decrypt


pcre2/NEWS
News about PCRE2 releases
-------------------------

Version 10.23 14-February-2017
------------------------------

1. ChangeLog has the details of a lot of bug fixes and tidies.

2. There has been a major re-factoring of the pcre2_compile.c file. Most syntax
checking is now done in the pre-pass that identifies capturing groups. This has
reduced the amount of duplication and made the code tidier. While doing this,
some minor bugs and Perl incompatibilities were fixed (see ChangeLog for
details.)

3. Back references are now permitted in lookbehind assertions when there are
no duplicated group numbers (that is, (?| has not been used), and, if the
reference is by name, there is only one group of that name. The referenced
group must, of course be of fixed length.

4. \g{+<number>} (e.g. \g{+2} ) is now supported. It is a "forward back
reference" and can be useful in repetitions (compare \g{-<number>} ). Perl does
not recognize this syntax.

5. pcre2grep now automatically expands its buffer up to a maximum set by
--max-buffer-size.

6. The -t option (grand total) has been added to pcre2grep.

7. A new function called pcre2_code_copy_with_tables() exists to copy a
compiled pattern along with a private copy of the character tables that is
uses.

8. A user supplied a number of patches to upgrade pcre2grep under Windows and
tidy the code.

9. Several updates have been made to pcre2test and test scripts (see
ChangeLog).


Version 10.22 29-July-2016
--------------------------

1. ChangeLog has the details of a number of bug fixes.

2. The POSIX wrapper function regcomp() did not used to support back references
and subroutine calls if called with the REG_NOSUB option. It now does.

3. A new function, pcre2_code_copy(), is added, to make a copy of a compiled
pattern.

4. Support for string callouts is added to pcre2grep.

5. Added the PCRE2_NO_JIT option to pcre2_match().

6. The pcre2_get_error_message() function now returns with a negative error
code if the error number it is given is unknown.

7. Several updates have been made to pcre2test and test scripts (see
ChangeLog).


Version 10.21 12-January-2016
-----------------------------

1. Many bugs have been fixed. A large number of them were provoked only by very
strange pattern input, and were discovered by fuzzers. Some others were
discovered by code auditing. See ChangeLog for details.

2. The Unicode tables have been updated to Unicode version 8.0.0.

3. For Perl compatibility in EBCDIC environments, ranges such as a-z in a
class, where both values are literal letters in the same case, omit the
non-letter EBCDIC code points within the range.

4. There have been a number of enhancements to the pcre2_substitute() function,
giving more flexibility to replacement facilities. It is now also possible to
cause the function to return the needed buffer size if the one given is too
small.

5. The PCRE2_ALT_VERBNAMES option causes the "name" parts of special verbs such
as (*THEN:name) to be processed for backslashes and to take note of
PCRE2_EXTENDED.

6. PCRE2_INFO_HASBACKSLASHC makes it possible for a client to find out if a
pattern uses \C, and --never-backslash-C makes it possible to compile a version
PCRE2 in which the use of \C is always forbidden.

7. A limit to the length of pattern that can be handled can now be set by
calling pcre2_set_max_pattern_length().

8. When matching an unanchored pattern, a match can be required to begin within
a given number of code units after the start of the subject by calling
pcre2_set_offset_limit().

9. The pcre2test program has been extended to test new facilities, and it can
now run the tests when LF on its own is not a valid newline sequence.

10. The RunTest script has also been updated to enable more tests to be run.

11. There have been some minor performance enhancements.


Version 10.20 30-June-2015
--------------------------

1. Callouts with string arguments and the pcre2_callout_enumerate() function
have been implemented.

2. The PCRE2_NEVER_BACKSLASH_C option, which locks out the use of \C, is added.

3. The PCRE2_ALT_CIRCUMFLEX option lets ^ match after a newline at the end of a
subject in multiline mode.

4. The way named subpatterns are handled has been refactored. The previous
approach had several bugs.

5. The handling of \c in EBCDIC environments has been changed to conform to the
perlebcdic document. This is an incompatible change.

6. Bugs have been mended, many of them discovered by fuzzers.


Version 10.10 06-March-2015
---------------------------

1. Serialization and de-serialization functions have been added to the API,
making it possible to save and restore sets of compiled patterns, though
restoration must be done in the same environment that was used for compilation.

2. The (*NO_JIT) feature has been added; this makes it possible for a pattern
creator to specify that JIT is not to be used.

3. A number of bugs have been fixed. In particular, bugs that caused building
on Windows using CMake to fail have been mended.


Version 10.00 05-January-2015
-----------------------------

Version 10.00 is the first release of PCRE2, a revised API for the PCRE
library. Changes prior to 10.00 are logged in the ChangeLog file for the old
API, up to item 20 for release 8.36. New programs are recommended to use the
new library. Programs that use the original (PCRE1) API will need changing
before linking with the new library.

****


server/core/maxpasswd.cc
/*
 * Copyright (c) 2016 MariaDB Corporation Ab
 * Copyright (c) 2023 MariaDB plc, Finnish Branch
 *
 * Use of this software is governed by the Business Source License included
 * in the LICENSE.TXT file and at www.mariadb.com/bsl11.
 *
 * Change Date: 2027-09-09
 *
 * On the date above, in accordance with the Business Source License, use
 * of this software will be governed by version 2 or later of the General
 * Public License.
 */

/**
 * @file maxpasswd.c  - Implementation of pasword encoding
 */

#include <maxscale/ccdefs.hh>

#include <cstdio>
#include <getopt.h>
#include <termios.h>
#include <unistd.h>
#include <maxbase/filesystem.hh>
#include <maxbase/log.hh>
#include <maxscale/key_manager.hh>
#include <maxscale/paths.hh>
#include <maxscale/utils.hh>
#include <iostream>

#include "internal/secrets.hh"

using std::cin;
using std::cout;
using std::endl;
using std::flush;
using std::string;

struct option options[] =
{
    {"help",        no_argument, nullptr, 'h'},
    {"decrypt",     no_argument, nullptr, 'd'},
    {"interactive", no_argument, nullptr, 'i'},
    {"keystore",    no_argument, nullptr, 'k'},
    {nullptr,       0,           nullptr, 0  }
};

void print_usage(const char* executable, const char* directory)
{
    const char msg[] =
        R"(Usage: %s [OPTION] [PATH] PASSWORD

Encrypt a MaxScale plaintext password using the encryption key in the key file
'%s'. The key file may be generated using the 'maxkeys'-utility.

  -h, --help         Display this help.
  -d, --decrypt      Decrypt an encrypted password instead.
  -i, --interactive  - If maxpasswd is reading from a pipe, it will read a line and
                       use that as the password.
                     - If maxpasswd is connected to a terminal console, it will prompt
                       for the password.
                     If '-i' is specified, a single argument is assumed to be the path
                     and two arguments is treated like an error.

  PATH      The key file directory (default: '%s').
  PASSWORD  The password to encrypt or decrypt. If --interactive is used, the password is read from stdin.
)";

    printf(msg, executable, SECRETS_FILENAME, directory);
}

bool read_password(string* pPassword)
{
    bool rv = false;
    string password;

    if (isatty(STDIN_FILENO))
    {
        struct termios tty;
        tcgetattr(STDIN_FILENO, &tty);

        bool echo = (tty.c_lflag & ECHO);
        if (echo)
        {
            tty.c_lflag &= ~ECHO;
            tcsetattr(STDIN_FILENO, TCSANOW, &tty);
        }

        cout << "Enter password : " << flush;
        string s1;
        std::getline(std::cin, s1);
        cout << endl;

        cout << "Repeat password: " << flush;
        string s2;
        std::getline(std::cin, s2);
        cout << endl;

        if (echo)
        {
            tty.c_lflag |= ECHO;
            tcsetattr(STDIN_FILENO, TCSANOW, &tty);
        }

        if (s1 == s2)
        {
            password = s1;
            rv = true;
        }
        else
        {
            cout << "Passwords are not identical." << endl;
        }
    }
    else
    {
        std::getline(std::cin, password);
        rv = true;
    }

    if (rv)
    {
        *pPassword = password;
    }

    return rv;
}


int main(int argc, char** argv)
{
    std::ios::sync_with_stdio();

    mxb::Log log(MXB_LOG_TARGET_STDOUT);
    const char* default_directory = mxs::datadir();

    enum class Mode
    {
        ENCRYPT,
        DECRYPT
    };

    auto mode = Mode::ENCRYPT;
    bool interactive = false;
    bool keystore = false;

    int c;
    while ((c = getopt_long(argc, argv, "hdi", options, NULL)) != -1)
    {
        switch (c)
        {
        case 'h':
            print_usage(argv[0], default_directory);
            return EXIT_SUCCESS;

        case 'd':
            mode = Mode::DECRYPT;
            break;

        case 'i':
            interactive = true;
            break;

        default:
            print_usage(argv[0], default_directory);
            return EXIT_FAILURE;
        }
    }

    string input;
    string path = default_directory;

    switch (argc - optind)
    {
    case 2:
        // Two args provided.
        path = argv[optind];
        if (!interactive)
        {
            input = argv[optind + 1];
        }
        else
        {
            print_usage(argv[0], default_directory);
            return EXIT_FAILURE;
        }
        break;

    case 1:
        // One arg provided.
        if (!interactive)
        {
            input = argv[optind];
        }
        else
        {
            path = argv[optind];
        }
        break;

    case 0:
        if (!interactive)
        {
            print_usage(argv[0], default_directory);
            return EXIT_FAILURE;
        }
        break;

    default:
        print_usage(argv[0], default_directory);
        return EXIT_FAILURE;
    }

    if (interactive)
    {
        if (!read_password(&input))
        {
            return EXIT_FAILURE;
        }
    }

    int rval = EXIT_FAILURE;

    string filepath = path;
    filepath.append("/").append(SECRETS_FILENAME);

    auto keydata = secrets_readkeys(filepath);
    if (keydata.ok)
    {
        bool encrypting = (mode == Mode::ENCRYPT);
        if (keydata.key.empty())
        {
            printf("Password encryption key file '%s' not found, cannot %s password.\n",
                   filepath.c_str(), encrypting ? "encrypt" : "decrypt");
        }
        else if (encrypting)
        {
            string encrypted = mxs::encrypt_password(keydata.key, input);
            if (!encrypted.empty())
            {
                printf("%s\n", encrypted.c_str());
                rval = EXIT_SUCCESS;
            }
            else
            {
                printf("Password encryption failed.\n");
            }
        }
        else
        {
            auto is_hex = std::all_of(input.begin(), input.end(), isxdigit);
            if (is_hex && input.length() % 2 == 0)
            {
                string decrypted = mxs::decrypt_password(keydata.key, input);
                if (!decrypted.empty())
                {
                    printf("%s\n", decrypted.c_str());
                    rval = EXIT_SUCCESS;
                }
                else
                {
                    printf("Password decryption failed.\n");
                }
            }
            else
            {
                printf("Input is not a valid hex-encoded encrypted password.\n");
            }
        }
    }
    else
    {
        printf("Could not read encryption key file '%s'.\n", filepath.c_str());
    }
    return rval;
}
